### DISCUSSION TITLE
Support Sign up ?

### DISCUSSION THREAD

**Discussion by Cajuteq** [authelia/authelia#5116] [Ideas]
It is sometimes useful to allow people to register on a service, but authelia current workflow forces user creation to be made beforehand.

So it could be useful to have a Registration process to allow new users to sign up. 

I think it should be made optional, maybe with "allow_sign_up": false by default.

It could work the same way password reset endpoints works, with a 3 steps process and an email validation.

I genuinely don't know how to link that with the different supported backends !

**Comment by james-d-elliott** [comment]
Hi there, sorry for the long delay. This is an item that I'm thinking about as a few people have asked.

**Reply by zylhuang** [reply]
It would be perfect if this feature can be added. May be considering an invitation code to allow people to register,  this can let my target audiences to register their own accounts, and also prevent irrelevant persons

**Reply by assassinliujie** [reply]
any progress? really want this feature

**Reply by james-d-elliott** [reply]
No, no one has taken up this feature in any capacity. No design has even been conceptualized. We would welcome someone to do come up with a technical design proposal.

We generally clearly communicate when we're working on something unless it's a small trivial thing. So this question never really needs to be asked.

**Reply by SpyrosMoux** [reply]
Hello! Any updates on this?

**Reply by chris2k20** [reply]
Any update on that. Would really love to have that feature, because that is the ONLY reason why I can not use it for my new project.. :/ 
The users should be able to register themselves. 

Thx for your awesome work <3

**Reply by itsdapi** [reply]
Yeah. Authelia is such a lightweight solution. The signup feature would be very nice <3

**Reply by Lagomorphem** [reply]
> Hi there, sorry for the long delay. This is an item that I'm thinking about as a few people have asked.

I’d like to offer a different perspective: I think Authelia isn’t the right place for a sign-up feature - it should stay focused on authentication. Registration simply isn’t part of that job, and adding it risks unnecessary complexity and bloat.

IMHO a better approach is to handle sign-up in a separate sibling project, like authelia/signelia (or similar). This keeps things flexible, lightweight, and secure by design:

* Users who don’t need sign-up simply don’t deploy the service - no code, no bugs. Or they want to use a third party solution or have hacked their own on-boarding web app, up to them...

* All other can simply add a route for e.g. authelia.example.com/register in their reverse proxy pointing to the registration web app. As it will use the same css and theme, nobody would notice the separation.

* Authentication can run with read-only LDAP access, improving security. Why provide read-write credentials to authelia/authelia for a registration once in a while (which is the reality, for most people)?

* Sign-up is more than hacking in username, password, and email - it may involve terms acceptance, additional data, verification, or group assignment. Handling all this in authelia/authelia would introduce a lot of complexity. A dedicated sign up and on-boarding service can manage it cleanly.

* Users could deploy authelia/_signelia_ without authelia/_authelia_, if they have an according usecase. 
 
This separation would keep authelia/authelia lightweight, secure-by-design and efficient (values emphasized by authelia.com), while still supporting full registration workflows when needed.

I perceive authelia as a module in my setup, and I do not expect it to become a full featured suite with all bells and whistles.. If I wanted that, I would have chosen another solution.

**Reply by Crowley723** [reply]
This is remarkably similar to what we have in mind.
Rather than splitting it to its own project, we are planning on defining different services and listeners.
```yaml
server:
  listeners:
    - name: authelia_portal
      address: 0.0.0.0:9091
      services:
        - portal
    - name: authelia_admin
      address: 0.0.0.0:9092
      services:
        - admin
    - name: authelia_signup
       address: 0.0.0.0:9093
       services:
         - signup
```
This allows admins to split up "services," bundle them together, disable services, etc.

This idea is independent from signup (and will probably be done before sign up is complete) and enables a lot of the benefits you mentioned across a number of the features we are interested in adding.

**Reply by Lagomorphem** [reply]
> This is remarkably similar to what we have in mind. Rather than splitting it to its own project, we are planning on defining different services and listeners.
> [...]
> This allows admins to split up "services," bundle them together, disable services, etc.

If there is a software having two (lets ignore admin panel here for obvious reasons) different roles, and each of these roles could perfectly work stand-alone, I'd recommend to split. What is the synergy having those two roles merged into one project? Why not merge another more or less random project as a "service", like nice profile pages for each user? This is as much related to authentication as registration and on-boarding is.

**Comment by joseph-finger** [comment]
How about copying the design from Mastodon ? They support different registrations policies, varying from open registration, over approval based registration, to invite links only. https://docs.joinmastodon.org/user/signup/#tos I think approval based registration could be interesting as, Authelia already provides several notifiers which could relay the application form the new user to a moderator

**Reply by itsdapi** [reply]
invite link only is great! Great for small personal usage

**Comment by hobgoblina** [comment]
Any progress on this yet? Would be great to be able to use Authelia for some nonprofit/activist groups that I maintain servers for, but we need the ability to generate invite links (ideally via API).

Authentik is the only out of the box solution I've identified that fits our needs, but Authentik comes with much heavier resource usage than Authelia... and the budget for these orgs is basically $0, so it hurts to have to pay for bigger VPS instances solely because of the auth service.

**Reply by Crowley723** [reply]
I've been working on user management, which should pave the way for registration, but to my knowledge, nobody is working on registration directly.

**Reply by james-d-elliott** [reply]
Yeah no one is, we don't have a clear picture of what people want yet either.

**Reply by huyz** [reply]
I want to simply be able to invite friends by email

**Reply by altmeista** [reply]
@james-d-elliott Just any kind of registration process would be really nice. Maybe you can begin with a normal registration with E-Mail validation like @Cajuteq mentioned. I think once that is implemented, it could be easy to enable registration via invitation links?

**Comment by mvdkleijn** [comment]
I'm evaluating Authelia for use in a larger service that we're setting up. We were looking at Authentik, but to be honest, their docs are not as good as I thought and Authentik is fairly heavy.

Support of self registration and some sort of user management backend would be required for our needs. 

I guess what I meant to ask is: are you lacking dev resources to work on this or are you lacking clarity on requirements / roadmap with regards to these kinds of features? (or both? :smile: )

**Reply by Crowley723** [reply]
There are a number of reasons, the main one is there are larger, more important features we need to prioritize. User management is in progress as are a number of other large features.

We would welcome a technical design proposal if anyone is willing to take that on and discuss that with us, either here on GitHub, Matrix, or Discord.

Regardless, its likely this feature will depend (at least in part) on the user management feature.

**Comment by slowmom** [comment]
Hmm. You don't have web-based admin interface.
You don't have user self sign up.
But then WHY ARE YOU NEEDED? Seriously?

**Reply by huyz** [reply]
Rude. Check yourself

**Reply by james-d-elliott** [reply]
Realistically it's quite apparent these are features that are both popular and we intend to add support for. However we take a very intentional approach to design, and these features both present a higher than average security risk not present with any of the existing features; you only have to look at other implementations in this space to see all of the abundant CVE's directly related to this.

We're a small team providing our spare time for free to develop this and related software. We take a security-first approach to development, and while that has not and will not ensure we avoid CVE's it will mitigate the potential of them. 

Rather than rush we'd prefer to take our time and do it right the first time. We do not and will never aim to be all things to all people, some people are going to want particular features we refuse to implement (this **_is not_** one of those), and maybe we're just not the right implementation for them, or you; which is fine. I think when other implementations try or appear to try to do this it's actually a huge red flag as to their security philosophy. Some people may not be patient enough to wait out our process which is also something we're comfortable with, it wont affect our architecture/design philosophy.

Maybe it's a misconception on my part but I don't get the feeling most other implementations do this. They are more focused on adding as many features as fast as possible. It may also be them trying to cut into this space, which is understandable.

There is a lot of evidence to say we're the most popular implementation within the sea of applications in our space; obviously I'm excluding corporate here since we're not specifically targeting corporate uses even though we're introducing traditionally corporate only features.

**Comment by quillaja** [comment]
My personal wish is for a registration method that does not require the use of passwords at all, with an email being optional. User creation, sign-in, and authentication would be based on passkey only. Optional email-based verification (via a link or OTP) for new account creation would be nice, and although I don't necessarily care about that myself I suspect many would want that. I don't personally care about account recovery if the person looses their passkeys, but again I imagine most other people would.

I also don't care at all about any sort of admin interface. Some sort of CLI command-based basic admin would be perfectly adequate, and as long as there is good documentation even just going into the db and doing manual sql-based manipulation is acceptable.

My personal use case mostly revolves around being able to offer simple web-apps with account-based data, but passwordless, single-factor (the passkey), and no required ties to the user's real-life identity (yes, very late 90s early 00s of me).

**Reply by james-d-elliott** [reply]
The registration implementation is going to require a unique username, unique email address which will be forcibly verified, a password, and some basic details about the users identity, most likely this will be their first name and last name.

I understand the password requirement may not be super popular, but it exists because of how LDAP servers are known to operate which prevents an empty password from being set either via policy, security requirements, or general inconsistency. While this could be done conditionally this is a very fragile design.

The customization as to what they can use to authenticate is detailed in [the roadmap](https://www.authelia.com/roadmap/active/granular-authorization/). In consideration for this we can ensure additional credential registrations are confirmed or registered during the registration process by allowing admins to define a specific rule for this i..e. require them to register a passkey and then only allow passkey auth by only having rules that use that policy, or a specific configuration option that reflects it.

It's likely it will be part of a CLI as well as the GUI.

**Reply by quillaja** [reply]
Thank you. The required passkey registration + passkey-only auth would work for me. If a password is required during registration, perhaps it could be reset with a long random string in an automated fashion after passkey registration, either by the admin (via some sort of script, possibly using the cli admin interface) or some sort of post-registration hook. That would effectively make the password "not present" without leaving any user-defined passwords intact.

**Comment by filips123** [comment]
I think there should be different types of user registration available, for example:

- Completely open registration: Users can register themselves by providing their details (email, username and password, optionally also names, avatar and other fields). If configured to be enabled, a verification email should be sent. Once verified, the user should be automatically assigned specific groups.

- Open registration with admin confirmations: Users themselves can request to be registered by providing their details, but their accounts stay disabled until administrators confirm/enable their accounts.

- Invite-only: Administrators can provide user details, except the password, and invite the user. The user should then receive a confirmation email, and once verified, they should be able to set their own password. This way, administrators don't need to know or set passwords for invited users.

- Manual registration: Same as above, except administrators also directly enter the password. Similar as user registration in LLDAP, except it would be quicker because it wouldn't require opening a separate website just to register a new user.

The first two options are mutually-exclusive, while the invitation feature and manual registration should work even when the registration is already open, to allow admins to quickly invite/add new users to the system.
